openvdb/openvdb/points/PointRasterizeFrustum.h

```diff
@@ -0,0 +1,1805 @@
+                    xyz.z() = z;
+
+                    Vec3R xyzWS = frustum.indexToWorld(xyz);
+
+                    double xDist = xyzWS.x() - positionWS.x();
+                    double yDist = xyzWS.y() - positionWS.y();
+                    double zDist = xyzWS.z() - positionWS.z();
+
+                    double distanceSqr = xDist*xDist+yDist*yDist+zDist*zDist;
+                    op(outXYZ, scale, attributeScale, distanceSqr, radiusWS*radiusWS);
+                }
+            }
+        }
+    }
+
+    void operator()(const PointLeafT& leaf, size_t) const
+    {
+        if (mInterrupter && mInterrupter->wasInterrupted()) {
+            tbb::task::self().cancel_group_execution();
+            return;
+        }
+
+        using AccessorT = tree::ValueAccessor<typename GridT::TreeType>;
+        using MaskAccessorT = tree::ValueAccessor<const MaskTree>;
+
+        const bool isBool = BoolTraits<ValueT>::IsBool;
+        const bool isFrustum = !mSettings.transform->isLinear();
+
+        const bool useRaytrace = mSettings.velocityMotionBlur || !mStaticCamera;
+        const bool useRadius = mSettings.useRadius;
+        const float radiusScale = mSettings.radiusScale;
+        const float radiusThreshold = std::sqrt(3.0f);
+
+        const float threshold = mSettings.threshold;
+        const bool scaleByVoxelVolume = !useRadius && mSettings.scaleByVoxelVolume;
+
+        const float shutterStartDt = mSettings.camera.shutterStart()/mSettings.framesPerSecond;
+        const float shutterEndDt = mSettings.camera.shutterEnd()/mSettings.framesPerSecond;
+        const int motionSteps = std::max(1, mSettings.motionSamples-1);
+
+        std::vector<Vec3d> motionPositions(motionSteps+1, Vec3d());
+        std::vector<Vec2s> frustumRadiusSizes(motionSteps+1, Vec2s());
+
+        const auto& pointsTransform = mGrid.constTransform();
+
+        const float voxelSize = static_cast<float>(mSettings.transform->voxelSize()[0]);
+
+        auto& grid = mCombinable.local();
+        auto& tree = grid.tree();
+        const auto& transform = *(mSettings.transform);
+
+        grid.setTransform(mSettings.transform->copy());
+
+        AccessorT valueAccessor(tree);
+
+        std::unique_ptr<MaskAccessorT> maskAccessor;
+        auto maskTree = mMask.getTreePtr();
+        if (maskTree)  maskAccessor.reset(new MaskAccessorT(*maskTree));
+
+        const CoordBBox* clipBBox = !mMask.clipBBox().empty() ? &mMask.clipBBox() : nullptr;
+
+        std::unique_ptr<AccessorT> weightAccessor;
+        if (mWeightCombinable) {
+            auto& weightGrid = mWeightCombinable->local();
+            auto& weightTree = weightGrid.tree();
+            weightAccessor.reset(new AccessorT(weightTree));
+        }
+
+        // create a temporary tree when rasterizing with radius and accumulate
+        // or weighted average methods
+
+        typename TreeT::Ptr tempTree;
+        typename TreeT::Ptr tempWeightTree;
```

Comment:
I'd try and cleanup and use of shared_ptr for temporaries - either just allocate them on the stack or use unique ptrs